Task3/EEL_task_3/src/main/java/com/epam/esm/giftcertificate/service/GiftCertificateService.java

```diff
@@ -3,165 +3,182 @@
-                    tagsToDelete = tagService.getTagsByNames(tagsToDeleteNames);
-                    giftCertificateRepository.deleteTagDependenciesForGiftCertificate(tagsToDelete.stream().map(Tag::getId).collect(Collectors.toList()), id);
-                    alreadyUsedTags.removeAll(tagsToDelete);
-                }
-                if (!tagsToUpdateNames.isEmpty()) {
-                    tagsToUpdate = tagsToUpdate.stream().filter(tag -> tagsToUpdateNames.contains(tag.getName())).collect(Collectors.toList());
-                    tagService.isTagsExistOrElseCreate(tagsToUpdate);
-                    tagsToUpdate = tagService.getTagsByNames(tagsToUpdateNames);
-                    giftCertificateRepository.createTagDependenciesForGiftCertificate(tagsToUpdate.stream().map(Tag::getId).collect(Collectors.toList()), id);
-                }
-                return true;
-
-            }
-            throw new ObjectIsInvalidException("Some tags are invalid");
-        }
+    public Page<GiftCertificateDTO> getAllGiftCertificates(Integer page, Integer size) {
+        PageRequest pageRequest = PageRequest.of(page, size);
+        return ParamsValidation.isListIsNotEmptyOrElseThrowNoSuchItem(giftCertificateRepository.findAll(pageRequest)).map(entityToDtoMapper::toGiftCertificateDTO);
     }
 
-    public GiftCertificate getCertificateById(long id) {
+    public GiftCertificateDTO getCertificateById(long id) {
 
-        GiftCertificate giftCertificate = giftCertificateRepository.getGiftCertificateByID(id);
-        if (giftCertificate == null) {
+        Optional<GiftCertificate> giftCertificate = giftCertificateRepository.findById(id);
+        if (giftCertificate.isEmpty()) {
             throw new NoSuchItemException("GiftCertificate with id = " + id + " doesn't exist");
         }
-        giftCertificate.setTags(getAllTagsByCertificate(giftCertificate));
-        return giftCertificate;
-
+        return entityToDtoMapper.toGiftCertificateDTO(giftCertificate.get());
     }
 
-    public List<GiftCertificate> setTagsInCertificates(List<GiftCertificate> gcs) {
-        for (GiftCertificate giftCertificate : gcs) {
-            giftCertificate.setTags(getAllTagsByCertificate(giftCertificate));
+
+    @Transactional
+    @Modifying
+    public GiftCertificateDTO updateCertificate(long id, GiftCertificateDTO updatedGiftCertificateDTO) {
+        GiftCertificate updatedGiftCertificate = entityToDtoMapper.toGiftCertificate(updatedGiftCertificateDTO);
+        Optional<GiftCertificate> gc = giftCertificateRepository.findById(id);
+        if (gc.isEmpty()) throw new NoSuchItemException("No gift certificate with id = " + id);
+
+        GiftCertificate giftCertificateFromDB = gc.get();
+        if (ParamsValidation.ifCertificateNeedsAnUpdate(updatedGiftCertificate, giftCertificateFromDB)) {
+            Map<String, String> updatingMap = ParamsValidation.isPatchCertificateValid(updatedGiftCertificate);
```

Comment:
I think that it's better to use json merge patch or similar approach for such logic